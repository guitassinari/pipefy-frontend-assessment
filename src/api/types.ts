export type Pipe = {
  name: string
  color: string
  cards_count: number
}