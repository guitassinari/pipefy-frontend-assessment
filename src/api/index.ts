import Queries from './queries'
import * as Types from './types'

export {
  Queries,
  Types
}