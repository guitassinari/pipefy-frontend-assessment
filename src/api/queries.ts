import { gql } from '@apollo/client'

const GET_ORGANIZATION = gql`
  query GetOrganization($id: ID!) {
    organization(id: $id) {
      id
      name
      pipes {
        id
        name
        color
        cards_count
      }
    }
  }
`

const GET_PIPE_CARDS = gql`
  query GetPipeCards($pipeId: ID!, $pageSize: Int, $startAtCursor: String) {
    cards(pipe_id: $pipeId, first: $pageSize, after: $startAtCursor) {
      pageInfo {
        endCursor
        hasNextPage
      }
      edges {
        cursor
        node {
          title
        }
      }
    }
  }
`

export default {
  GET_ORGANIZATION,
  GET_PIPE_CARDS
}