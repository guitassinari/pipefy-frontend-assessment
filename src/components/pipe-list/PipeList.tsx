import React, { FC } from 'react';
import {  Types } from 'api'
import PipeTile from 'components/pipe-tile';
import { PipeListContainer } from './PipeList.styles'

interface PipeListProps {
  pipes: Types.Pipe[]
}

const PipeList: FC<PipeListProps> = ({ pipes }) => (
  <PipeListContainer role="grid">
    {pipes.map((pipe: Types.Pipe, index: number) => (
      <PipeTile role="gridcell" key={index} pipe={pipe} />
    ))}
  </PipeListContainer>
)


export default PipeList
