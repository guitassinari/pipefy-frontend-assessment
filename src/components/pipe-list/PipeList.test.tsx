import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PipeList from '.'
import { Types } from 'api';

describe('PipeList', () => {
  describe('When receives a valid list of pipes', () => {
    const pipes: Types.Pipe[] = [{
      name: 'A pipe',
      color: 'blue',
      cards_count: 10,
      id: 1
    }]

    it('matches snapshot', () => {
      const component = render(<PipeList pipes={pipes}/>)

      expect(component.asFragment()).toMatchSnapshot()
    })

    it('shows a list of the organization pipes', async () => {
      const { getByText } = render(<PipeList pipes={pipes} />)

      pipes.forEach(pipe => {
        expect(getByText(pipe.name)).toBeInTheDocument()
      })
    })

    describe('when receives onClick prop', () => {
      describe('and a pipe is clicked', () => {
        it('calls the onClick callback with that pipe', () => {
          const onClickCallback = jest.fn()
          const { getByText } = render(<PipeList onClick={onClickCallback} pipes={pipes} />)

          fireEvent.click(getByText(pipes[0].name))

          expect(onClickCallback).toHaveBeenCalledWith(pipes[0])
        })
      })
    })
  })
})
