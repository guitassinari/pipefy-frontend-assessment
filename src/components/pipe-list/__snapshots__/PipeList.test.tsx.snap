// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PipeList When receives a valid list of pipes matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="sc-gsDKAQ fqXist"
  >
    <div
      class="sc-bdvvtL fFSuzC"
    >
      <div>
        A pipe
      </div>
    </div>
  </ul>
</DocumentFragment>
`;
